File: launcher/MonitorCanvas.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Threading;
using System.Windows.Forms;

internal static class MonitorCanvasLauncher
{
    private const string ApplicationUrl = "http://127.0.0.1:8765";
    private const string HealthUrl = ApplicationUrl + "/api/monitors";
    private const string MutexName = @"Local\MonitorCanvas.LocalService";

    [STAThread]
    private static void Main()
    {
        if (IsServiceReady())
        {
            OfferToOpenRunningApplication();
            return;
        }

        bool ownsMutex;
        using (var singleInstance = new Mutex(true, MutexName, out ownsMutex))
        {
            if (!ownsMutex)
            {
                WaitUntilReady();
                OfferToOpenRunningApplication();
                return;
            }

            if (IsServiceReady())
            {
                OfferToOpenRunningApplication();
                return;
            }

            StartApplication();
        }
    }

    private static void StartApplication()
    {
        string directory = AppDomain.CurrentDomain.BaseDirectory;
        string scriptPath = Path.Combine(directory, "server.ps1");

        if (!File.Exists(scriptPath))
        {
            MessageBox.Show(
                "Die Datei server.ps1 wurde nicht gefunden.",
                "MonitorCanvas",
                MessageBoxButtons.OK,
                MessageBoxIcon.Error
            );
            return;
        }

        try
        {
            var startInfo = new ProcessStartInfo
            {
                FileName = "powershell.exe",
                Arguments =
                    "-NoLogo -NoProfile -ExecutionPolicy Bypass -WindowStyle Hidden " +
                    "-File \"" + scriptPath + "\" -NoBrowser",
                WorkingDirectory = directory,
                UseShellExecute = false,
                CreateNoWindow = true,
                WindowStyle = ProcessWindowStyle.Hidden
            };

            Process.Start(startInfo);
            WaitUntilReady();
            OpenApplication();
        }
        catch (Exception error)
        {
            MessageBox.Show(
                "MonitorCanvas konnte nicht gestartet werden.\n\n" + error.Message,
                "MonitorCanvas",
                MessageBoxButtons.OK,
                MessageBoxIcon.Error
            );
        }
    }

    private static void WaitUntilReady()
    {
        for (int attempt = 0; attempt < 40; attempt++)
        {
            if (IsServiceReady())
            {
                return;
            }
            Thread.Sleep(250);
        }

        throw new InvalidOperationException(
            "Der lokale Dienst antwortet nicht. Bitte MonitorCanvas erneut starten."
        );
    }

    private static bool IsServiceReady()
    {
        try
        {
            var request = WebRequest.Create(HealthUrl);
            request.Timeout = 500;
            using (request.GetResponse())
            {
                return true;
            }
        }
        catch
        {
            return false;
        }
    }

    private static void OfferToOpenRunningApplication()
    {
        DialogResult result = MessageBox.Show(
            "MonitorCanvas läuft bereits.\n\nSoll die Anwendung im Browser geöffnet werden?",
            "MonitorCanvas",
            MessageBoxButtons.YesNo,
            MessageBoxIcon.Information
        );
        if (result == DialogResult.Yes)
        {
            OpenApplication();
        }
    }

    private static void OpenApplication()
    {
        Process.Start(new ProcessStartInfo
        {
            FileName = ApplicationUrl,
            UseShellExecute = true
        });
    }
}

File: launcher/MonitorCanvas.cpp
#define UNICODE
#define _UNICODE

#include <windows.h>
#include <shellapi.h>
#include <string>

int WINAPI wWinMain(HINSTANCE, HINSTANCE, PWSTR, int)
{
    wchar_t executablePath[MAX_PATH];
    if (!GetModuleFileNameW(nullptr, executablePath, MAX_PATH)) {
        MessageBoxW(nullptr, L"Der Programmordner konnte nicht ermittelt werden.",
                    L"MonitorCanvas", MB_OK | MB_ICONERROR);
        return 1;
    }

    std::wstring directory(executablePath);
    const size_t separator = directory.find_last_of(L"\\/");
    if (separator != std::wstring::npos) {
        directory.resize(separator);
    }

    const std::wstring scriptPath = directory + L"\\server.ps1";
    const std::wstring command =
        L"powershell.exe -NoLogo -NoProfile -ExecutionPolicy Bypass -WindowStyle Hidden "
        L"-File \"" + scriptPath + L"\" -NoBrowser";

    STARTUPINFOW startupInfo{};
    startupInfo.cb = sizeof(startupInfo);
    startupInfo.dwFlags = STARTF_USESHOWWINDOW;
    startupInfo.wShowWindow = SW_HIDE;

    PROCESS_INFORMATION processInfo{};
    std::wstring mutableCommand = command;
    const BOOL started = CreateProcessW(
        nullptr,
        mutableCommand.data(),
        nullptr,
        nullptr,
        FALSE,
        CREATE_NO_WINDOW,
        nullptr,
        directory.c_str(),
        &startupInfo,
        &processInfo
    );

    if (!started) {
        MessageBoxW(nullptr,
                    L"Der lokale MonitorCanvas-Dienst konnte nicht gestartet werden.",
                    L"MonitorCanvas", MB_OK | MB_ICONERROR);
        return 1;
    }

    CloseHandle(processInfo.hThread);
    CloseHandle(processInfo.hProcess);
    Sleep(1800);
    ShellExecuteW(nullptr, L"open", L"http://127.0.0.1:8765", nullptr, nullptr, SW_SHOWNORMAL);
    return 0;
}
